File: Arrays & Hashing/13. Roman to Integer/Solution.java
import java.util.*;

class Solution {
    public int romanToInt(String s) {
        // N : len(s)
        // Time Complexity: O(N)
        // Space Complexity: O(1)
        int n = s.length();
        HashMap<Character, Integer> romanMap = new HashMap<>(Map.of(
                'I', 1,
                'V', 5,
                'X', 10,
                'L', 50,
                'C', 100,
                'D', 500,
                'M', 1000
        ));
        int num = 0;
        for(int i=n-1; i>=0; i--){
            if (i < n - 1 && romanMap.get(s.charAt(i)) < romanMap.get(s.charAt(i + 1))){
                num -= romanMap.get(s.charAt(i));
            } else{
                num += romanMap.get(s.charAt(i));
            } 
        }
        return num;
    }
}

File: Arrays & Hashing/13. Roman to Integer/Solution.py
class Solution:
    def romanToInt(self, s: str) -> int:
        # N : len(s)
        # Time Complexity: O(N)
        # Space Complexity: O(1)
        n = len(s)
        romanNumeralMap = {
            'I': 1, 
            'V': 5, 
            'X': 10, 
            'L': 50,
            'C': 100, 
            'D': 500, 
            'M': 1000
        }
        num = 0
        for i in range(n-1, -1, -1):
            if i < n - 1 and romanNumeralMap[s[i]] < romanNumeralMap[s[i + 1]]:
                num -= romanNumeralMap[s[i]]
            else:
                num += romanNumeralMap[s[i]]
        return num